File: runtime/src/main/java/com/siimkinks/sqlitemagic/UnaryExpr.java
package com.siimkinks.sqlitemagic;

import android.support.annotation.NonNull;

import com.siimkinks.sqlitemagic.internal.SimpleArrayMap;

import java.util.ArrayList;
import java.util.LinkedList;

final class UnaryExpr extends Expr {
  @NonNull
  final Expr expr;

  UnaryExpr(@NonNull String op, @NonNull Expr expr) {
    //noinspection ConstantConditions
    super(null, op); // null is ok here, we override all calls related to column
    this.expr = expr;
  }

  @Override
  void addArgs(@NonNull ArrayList<String> args) {
    expr.addArgs(args);
  }

  @Override
  void appendToSql(@NonNull StringBuilder sb) {
    sb.append(op);
    sb.append('(');
    expr.appendToSql(sb);
    sb.append(')');
  }

  @Override
  void appendToSql(@NonNull StringBuilder sb, @NonNull SimpleArrayMap<String, LinkedList<String>> systemRenamedTables) {
    sb.append(op);
    sb.append('(');
    expr.appendToSql(sb, systemRenamedTables);
    sb.append(')');
  }

  @Override
  boolean containsColumn(@NonNull Column<?, ?, ?, ?, ?> column) {
    return expr.containsColumn(column);
  }
}

File: runtime/src/main/kotlin/com/siimkinks/sqlitemagic/UnaryExpr.kt
package com.siimkinks.sqlitemagic

import com.siimkinks.sqlitemagic.internal.SimpleArrayMap
import java.util.LinkedList

internal class UnaryExpr(
  op: String,
  private val expr: Expr
) : Expr(
  column = null,
  op = op
) {
  override fun addArgs(args: ArrayList<String?>) {
    expr.addArgs(args)
  }

  override fun appendToSql(sb: StringBuilder) {
    sb.appendUnary {
      expr.appendToSql(sb)
    }
  }

  override fun appendToSql(
    sb: StringBuilder,
    systemRenamedTables: SimpleArrayMap<String, LinkedList<String>>
  ) {
    sb.appendUnary {
      expr.appendToSql(sb, systemRenamedTables)
    }
  }

  private inline fun StringBuilder.appendUnary(
    appendExpr: () -> Unit
  ) {
    append(op)
    append('(')
    appendExpr()
    append(')')
  }

  override fun containsColumn(column: Column<*, *, *, *, *>) = expr.containsColumn(column)
}
